domain/secretbackend/state/state.go

```diff
@@ -512,51 +521,86 @@ WHERE uuid = $M.uuid`, sqlair.M{}, SecretBackendRotationRow{})
+	db, err := s.DB()
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	var modelBackendDetails domainsecretbackend.ModelSecretBackend
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		modelBackendDetails, err = s.getModelSecretBackendDetails(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return "", "", domain.CoerceError(err)
+	}
+	return modelBackendDetails.SecretBackendName, modelBackendDetails.Type, nil
+}
+
 // SetModelSecretBackend sets the secret backend for the given model.
-func (s *State) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, backendName string) error {
+func (s *State) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, getBackendName func(coremodel.ModelType) (string, error)) error {
 	db, err := s.DB()
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	modelBackendUpsert := `
-INSERT INTO model_secret_backend
-	(model_uuid, secret_backend_uuid)
-VALUES ($M.model_uuid, $SecretBackendRow.uuid)
-ON CONFLICT (model_uuid) DO UPDATE SET
-	secret_backend_uuid = EXCLUDED.secret_backend_uuid`
-	modelBackendUpsertStmt, err := s.Prepare(modelBackendUpsert, sqlair.M{}, SecretBackendRow{})
+	secretBackendSelectQ := `
+SELECT b.uuid AS &ModelSecretBackend.secret_backend_uuid
+FROM   secret_backend b
+WHERE  b.name = $ModelSecretBackend.secret_backend_name
+`
+	backendInfo := ModelSecretBackend{ID: modelUUID}
+	secretBackendSelectStmt, err := s.Prepare(secretBackendSelectQ, backendInfo)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	q := `
-SELECT b.uuid AS &SecretBackendRow.uuid
-FROM secret_backend b
-WHERE b.name = $SecretBackendRow.name`
-	stmt, err := s.Prepare(q, SecretBackendRow{})
+	modelBackendUpdate := `
+UPDATE model_secret_backend
+SET    secret_backend_uuid = $ModelSecretBackend.secret_backend_uuid
+WHERE  model_uuid = $ModelSecretBackend.uuid`
+	modelBackendUpdateStmt, err := s.Prepare(modelBackendUpdate, backendInfo)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		backend := SecretBackendRow{Name: backendName}
+		modelBackendDetails, err := s.getModelSecretBackendDetails(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		backendInfo.SecretBackendName, err = getBackendName(modelBackendDetails.Type)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if backendInfo.SecretBackendName == modelBackendDetails.SecretBackendName {
+			// Nothing to update.
+			return nil
+		}
 
-		var result SecretBackendRow
-		err = tx.Query(ctx, stmt, backend).Get(&result)
+		err = tx.Query(ctx, secretBackendSelectStmt, backendInfo).Get(&backendInfo)
 		if errors.Is(err, sql.ErrNoRows) {
-			return fmt.Errorf("%w: %q", secretbackenderrors.NotFound, backendName)
+			return fmt.Errorf("%w: %q", secretbackenderrors.NotFound, backendInfo.SecretBackendName)
```

Comment:
can we make this `fmt.Errorf("cannot set model %q secret backend to %q: %w", modelID, backendName, backendInfo.SecretBackendName)`